File: First negative integer in every window of size k - GFG/first-negative-integer-in-every-window-of-size-k.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

vector<long long> printFirstNegativeInteger(long long int arr[],
                                             long long int n, long long int k);

// Driver program to test above functions
int main() {
    long long int t, i;
    cin >> t;
    while (t--) {
        long long int n;
        cin >> n;
        long long int arr[n];
        for (i = 0; i < n; i++) {
            cin >> arr[i];
        }
        long long int k;
        cin >> k;

        vector<long long> ans = printFirstNegativeInteger(arr, n, k);
        for (auto it : ans) cout << it << " ";
        cout << endl;
    }
    return 0;
}

// } Driver Code Ends


vector<long long> printFirstNegativeInteger(long long int A[],
                                             long long int N, long long int K) {
    vector<long long> ans;
    long long i=0;
    long long j=0;
    queue<long long> q;
    while(j<N){
        if(A[j]<0){
            q.push(A[j]);
        }
        if(j-i+1<K){
            j++;
        }
        else if(j-i+1==K){
            if(q.empty()){
                ans.push_back(0);
            }
            else{
                ans.push_back(q.front());    
            }
            if(A[i]<0){
                q.pop();
            }
            i++;
            j++;
        }
    }
    return ans;
 }

File: First negative integer in every window of size k - GFG/first-negative-integer-in-every-window-of-size-k.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;

class GFG {
	public static void main(String[] args) throws IOException
	{
	        BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));
        int t =
            Integer.parseInt(br.readLine().trim()); // Inputting the testcases
        while(t-->0)
        {
            int n = Integer.parseInt(br.readLine().trim());
            long a[] = new long[(int)(n)];
            // long getAnswer[] = new long[(int)(n)];
            String inputLine[] = br.readLine().trim().split(" ");
            for (int i = 0; i < n; i++) {
                a[i] = Long.parseLong(inputLine[i]);
            }
            int k = Integer.parseInt(br.readLine().trim());
            
            Compute obj = new Compute();
            long answer[] = obj.printFirstNegativeInteger(a, n, k);
            int sz = answer.length;
            
            StringBuilder output = new StringBuilder();
            for(int i=0;i<sz;i++)
                output.append(answer[i]+" ");
            System.out.println(output);
            
        }
	}
}


// } Driver Code Ends

class Compute {
    
    public long[] printFirstNegativeInteger(long nums[], int n, int k)
    {
        long result[] = new long[n-k+1];
        Queue<Long> q = new LinkedList<>();
        int j=0;
        for(int i=0; i<k; i++){
            if(nums[i] < 0){
                q.add(nums[i]);
            }
        }
        if(q.isEmpty()){
            result[j++] = 0;
        }
        else{
            result[j++] = q.peek();
        }
        for(int i=k; i<n; i++){
            if(nums[i] < 0){
                q.add(nums[i]);
            }
            if(nums[i-k] < 0){
                q.poll();
            }
            if(q.isEmpty()){
                result[j++] = 0;
            }
            else{
                result[j++] = q.peek();
            }
        }
        return result;
    }
}